diff -rN -u mod_authz_ldap-0.30.old/module/auth.c mod_authz_ldap-0.30/module/auth.c
--- mod_authz_ldap-0.30.old/module/auth.c	2012-04-26 14:54:42.000000000 +0200
+++ mod_authz_ldap-0.30/module/auth.c	2012-04-26 14:54:43.000000000 +0200
@@ -92,10 +92,9 @@
 	/* password, so we get them here quite early. This should never	*/
 	/* return NULL, as we go to this function only when doing	*/
 	/* basic authentication						*/
-	AUTHZ_RDEBUG("[%d] authz_ldap_authenticate(user = %s, pw = %s, dn = %s)",
+	AUTHZ_RDEBUG("[%d] authz_ldap_authenticate(user = %s, dn = %s)",
 		(int)getpid(),
 		(USER(r)) ? USER(r) : "(null)",
-		(pw) ? pw : "(null)",
 		(userdn) ? userdn : "(null)");
 	ap_get_basic_auth_pw(r, (const char **)&pw);
 
diff -rN -u mod_authz_ldap-0.30.old/module/utilities.c mod_authz_ldap-0.30/module/utilities.c
--- mod_authz_ldap-0.30.old/module/utilities.c	2012-04-26 14:54:42.000000000 +0200
+++ mod_authz_ldap-0.30/module/utilities.c	2012-04-26 14:54:43.000000000 +0200
@@ -114,8 +114,8 @@
 		if (sec->loglevel >= APLOG_EMERG) {
 			ap_log_rerror(APLOG_MARK, APLOG_EMERG, 0, r,
 				"cannot bind to "
-				"[%d] LDAP Server as %s/%s: %d = '%s'", (int)getpid(), 
-				sec->binddn, sec->bindpw, rc, ldap_err2string(rc));
+				"[%d] LDAP Server as %s: %d = '%s'", (int)getpid(), 
+				sec->binddn, rc, ldap_err2string(rc));
 		}
 		return -1;
 	}
